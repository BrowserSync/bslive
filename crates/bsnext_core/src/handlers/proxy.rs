use anyhow::Context;
use axum::body::Body;
use axum::extract::Request;
use axum::response::{IntoResponse, Response};
use axum::routing::any;
use axum::Extension;
use bsnext_guards::OuterUri;
use http::uri::{Parts, PathAndQuery};
use http::{HeaderName, HeaderValue, StatusCode, Uri};
use hyper_tls::HttpsConnector;
use hyper_util::client::legacy::connect::HttpConnector;
use hyper_util::client::legacy::Client;
use std::collections::BTreeMap;
use std::str::FromStr;
use tower::ServiceExt;
use tracing::{trace_span, Instrument};

#[derive(Debug, Clone)]
pub struct ProxyConfig {
    pub target: String,
    pub path: String,
    pub headers: BTreeMap<String, String>,
    pub rewrite_kind: RewriteKind,
}

#[derive(Debug, Clone)]
pub enum RewriteKind {
    Alias,
    Nested,
}

impl From<Option<bool>> for RewriteKind {
    fn from(value: Option<bool>) -> Self {
        match value {
            None | Some(true) => RewriteKind::Nested,
            Some(false) => RewriteKind::Alias,
        }
    }
}

// Make our own error that wraps `anyhow::Error`.
pub struct AnyAppError(anyhow::Error);

// Tell axum how to convert `AppError` into a response.
impl IntoResponse for AnyAppError {
    fn into_response(self) -> Response {
        (
            StatusCode::INTERNAL_SERVER_ERROR,
            format!("Something went wrong: {}", self.0),
        )
            .into_response()
    }
}
impl<E> From<E> for AnyAppError
where
    E: Into<anyhow::Error>,
{
    fn from(err: E) -> Self {
        Self(err.into())
    }
}

pub async fn proxy_handler(
    Extension(config): Extension<ProxyConfig>,
    Extension(OuterUri(outer_uri)): Extension<OuterUri>,
    uri: Uri,
    req: Request,
) -> Result<Response, AnyAppError> {
    let span = trace_span!("proxy_handler");
    let _g = span.enter();

    tracing::trace!(outer_uri = ?outer_uri);
    tracing::trace!(uri = ?uri);
    tracing::trace!(config.path = ?config.path);
    tracing::trace!(config.target = ?config.target);
    tracing::trace!(config.headers = ?config.headers);
    tracing::trace!(config.rewrite_kind = ?config.rewrite_kind);

    let target = config.target.clone();
    let target_uri = Uri::try_from(&target)?;

    tracing::trace!(target = ?target);
    tracing::trace!(target.uri = ?target_uri);
    tracing::trace!(target.uri.authority = ?target_uri.authority());
    tracing::trace!(target.uri.path_and_query = ?target_uri.path_and_query());

    let args = IntoTarget {
        target_uri: &target_uri,
        outer_uri: &outer_uri,
        uri: &uri,
        config: &config,
    };
    let parsed = into_target_uri(args)?;
    tracing::trace!(outgoing.uri = %parsed);

    let (parts, body) = req.into_parts();

    let mut req = Request::from_parts(parts, body);

    let target = match (parsed.host(), parsed.port()) {
        (Some(host), Some(port)) => format!("{host}:{port}"),
        (Some(host), None) => host.to_owned(),
        _ => unreachable!("could not extract `host` from url"),
    };

    *req.uri_mut() = parsed;
    let host_header_value = HeaderValue::from_str(&target)?;
    req.headers_mut().insert("host", host_header_value);
    req.headers_mut().remove("referer");

    for (k, v) in config.headers {
        match (
            HeaderName::from_bytes(k.as_bytes()),
            HeaderValue::from_bytes(v.as_bytes()),
        ) {
            (Ok(hn), Ok(hv)) => {
                tracing::trace!(?hn, ?hv, "add cookie to outgoing request");
                req.headers_mut().insert(hn, hv);
            }
            _ => {
                // noop
            }
        }
    }

    let sv2 = any(serve_one_proxy_req);
    Ok(sv2
        .oneshot(req)
        .instrument(span.clone())
        .await
        .into_response())
}

#[tracing::instrument(skip_all)]
async fn serve_one_proxy_req(req: Request) -> Response {
    tracing::trace!("serve_one_proxy_req {}", req.uri().to_string());
    let client = {
        req.extensions()
            .get::<Client<HttpsConnector<HttpConnector>, Body>>()
            .expect("must have a client, move this to an extractor?")
    };
    tracing::trace!(req.headers = ?req.headers());
    tracing::trace!(req.method = ?req.method());
    let client_c = client.clone();
    let resp = client_c
        .request(req)
        .await
        .map_err(|_| StatusCode::BAD_REQUEST)
        .into_response();
    tracing::trace!(resp.status = resp.status().as_u16());
    tracing::trace!(resp.headers = ?resp.headers());
    resp
}

struct IntoTarget<'a> {
    target_uri: &'a Uri,
    outer_uri: &'a Uri,
    uri: &'a Uri,
    config: &'a ProxyConfig,
}

fn into_target_uri(
    IntoTarget {
        target_uri,
        outer_uri,
        uri,
        config,
    }: IntoTarget,
) -> anyhow::Result<Uri> {
    let uri_src = match config.rewrite_kind {
        RewriteKind::Alias => &outer_uri,
        RewriteKind::Nested => &uri,
    };

    let req_query = uri_src.path_and_query().and_then(|x| x.query());
    let target = target_uri.path_and_query().filter(|x| x.path() != "/");
    let src = uri_src.path_and_query().filter(|x| x.path() != "/");

    let next_pq = match (target, src) {
        (Some(target), Some(req)) => {
            let p1 = target.path();
            let p2 = req.path();
            let next = match req_query {
                None => format!("{p1}{p2}"),
                Some(q) => format!("{p1}{p2}?{q}"),
            };
            let v =
                PathAndQuery::from_str(&next).unwrap_or(uri_src.path_and_query().unwrap().clone());
            v
        }
        (Some(target_only), None) => {
            let path = target_only.path();

            let next = match req_query {
                None => path,
                Some(q) => &format!("{path}?{q}"),
            };
            let v =
                PathAndQuery::from_str(next).unwrap_or(uri_src.path_and_query().unwrap().clone());
            v
        }
        (None, Some(req_only)) => req_only.to_owned(),
        (None, None) => uri_src.path_and_query().unwrap().clone(),
    };

    let mut next_parts = Parts::default();
    if let Some(scheme) = target_uri.scheme() {
        next_parts.scheme = Some(scheme.to_owned())
    }
    if let Some(auth) = target_uri.authority() {
        next_parts.authority = Some(auth.to_owned());
    }

    next_parts.path_and_query = Some(next_pq);

    Uri::from_parts(next_parts).context("tried to parse")
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::str::FromStr;

    struct TestCase {
        name: &'static str,
        target: &'static str,
        outer: &'static str,
        uri: &'static str,
        rewrite: RewriteKind,
        expect: &'static str,
    }

    #[test]
    fn test_into_target_uri() {
        let test_cases = vec![
            TestCase {
                name: "basic",
                target: "http://example.com",
                outer: "/",
                uri: "/",
                rewrite: RewriteKind::Alias,
                expect: "http://example.com/",
            },
            TestCase {
                name: "basic nested alias - the `/api` section is preserved here",
                target: "http://example.com",
                outer: "/api/abc",
                uri: "/abc",
                rewrite: RewriteKind::Alias,
                expect: "http://example.com/api/abc",
            },
            TestCase {
                name: "basic nested rewrite - the `/api` prefixed is dropped here",
                target: "http://example.com",
                outer: "/api/abc",
                uri: "/abc",
                rewrite: RewriteKind::Nested,
                expect: "http://example.com/abc",
            },
            TestCase {
                name: "basic nested rewrite",
                target: "http://example.com/api",
                outer: "http://localhost:3000/test",
                uri: "/users/123",
                rewrite: RewriteKind::Nested,
                expect: "http://example.com/api/users/123",
            },
            TestCase {
                name: "basic alias rewrite",
                target: "http://example.com/api",
                outer: "http://localhost:3000/test/users/123",
                uri: "/users/123",
                rewrite: RewriteKind::Alias,
                expect: "http://example.com/api/test/users/123",
            },
            TestCase {
                name: "with query params",
                target: "http://example.com/api",
                outer: "http://localhost:3000/test?foo=bar",
                uri: "/users/123?foo=bar",
                rewrite: RewriteKind::Nested,
                expect: "http://example.com/api/users/123?foo=bar",
            },
        ];

        for tc in test_cases {
            let target_uri = Uri::from_str(tc.target).unwrap();
            let outer_uri = Uri::from_str(tc.outer).unwrap();
            let uri = Uri::from_str(tc.uri).unwrap();
            let config = ProxyConfig {
                target: tc.target.to_string(),
                path: String::new(),
                headers: BTreeMap::new(),
                rewrite_kind: tc.rewrite,
            };

            let args = IntoTarget {
                target_uri: &target_uri,
                outer_uri: &outer_uri,
                uri: &uri,
                config: &config,
            };

            let result = into_target_uri(args).unwrap();
            assert_eq!(
                result.to_string(),
                tc.expect,
                "failed test case: {}",
                tc.name
            );
        }
    }
}
