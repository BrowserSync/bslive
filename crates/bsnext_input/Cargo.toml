[package]
name = "bsnext_input"
version = "0.2.3"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
bsnext_resp = { path = "../bsnext_resp" }
bsnext_tracing = { path = "../bsnext_tracing" }
bsnext_guards = { path = "../bsnext_guards" }

miette = { workspace = true }

random_word = { workspace = true }
matchit = { workspace = true }
mime_guess = { workspace = true }
serde = { workspace = true }
clap = { workspace = true }
anyhow = { workspace = true }
serde_yaml = { workspace = true }
serde_json = { workspace = true }
toml = { workspace = true }
thiserror = { workspace = true }

[dev-dependencies]
insta = { workspace = true }
