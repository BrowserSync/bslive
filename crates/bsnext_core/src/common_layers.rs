use axum::extract::Request;
use axum::middleware::Next;
use axum::{middleware, Extension, Router};
use http::{HeaderName, HeaderValue};
use std::cmp::PartialEq;
use std::convert::Infallible;
use std::time::Duration;

use bsnext_input::route::{CorsOpts, DelayKind, DelayOpts, Route};
use bsnext_resp::injector_guard::InjectorGuard;
use bsnext_resp::{response_modifications_layer, InjectHandling};
use tokio::time::sleep;
use tower_http::cors::CorsLayer;
use tower_http::decompression::DecompressionLayer;
use tower_http::set_header;

#[derive(Debug, PartialEq)]
pub enum Handling {
    Proxy,
    Raw,
    Dir,
}

pub fn add_route_layers(app: Router, handling: Handling, route: &Route, req: &Request) -> Router {
    let mut app = app;

    if route
        .cors_opts
        .as_ref()
        .is_some_and(|v| *v == CorsOpts::Cors(true))
    {
        tracing::trace!(to = route.path, "adding permissive cors");
        app = app.layer(CorsLayer::permissive());
    }

    if let Some(DelayOpts::Delay(DelayKind::Ms(ms))) = route.delay_opts.as_ref() {
        tracing::trace!(to = route.path, ?ms, "adding a delay");
        let ms = *ms;
        app = app.layer(middleware::from_fn(
            move |req: Request, next: Next| async move {
                let res = next.run(req).await;
                sleep(Duration::from_millis(ms)).await;
                Ok::<_, Infallible>(res)
            },
        ));
    }

    if let Some(headers) = route.headers.as_ref() {
        for (k, v) in headers {
            let hn = HeaderName::from_bytes(k.as_bytes());
            let hv = HeaderValue::from_bytes(v.as_bytes());
            match (hn, hv) {
                (Ok(n), Ok(v)) => {
                    app = app.layer(set_header::SetResponseHeaderLayer::overriding(n, v));
                }
                (Ok(_), Err(_e)) => {
                    tracing::error!("invalid header value `{}`", v)
                }
                (Err(_e), Ok(_)) => {
                    tracing::error!("invalid header name `{}`", k)
                }
                (Err(_e), Err(_e2)) => {
                    tracing::error!("invalid header name AND value `{}:{}`", k, v)
                }
            }
        }
    }

    let injections = route.inject_opts.injections();
    let might_inject = injections.iter().any(|inj| inj.accept_req(req));

    if might_inject {
        if handling == Handling::Proxy {
            app = app.layer(DecompressionLayer::new())
        }
        app = app
            .layer(middleware::from_fn(response_modifications_layer))
            .layer(Extension(InjectHandling { items: injections }));
    }

    // if route.opts.as_ref().is_some_and(|v| v.buff) {
    // app = app.layer(middleware::from_fn(print_request_response));
    // }

    app
}
