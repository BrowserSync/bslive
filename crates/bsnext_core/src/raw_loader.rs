use std::convert::Infallible;

use axum::extract::{Request, State};
use axum::middleware::Next;
use axum::response::{Html, IntoResponse, Response, Sse};
use axum::routing::any;
use axum::{middleware, Json, Router};
use http::header::CONTENT_TYPE;

use crate::meta::MetaData;
use crate::server::state::ServerState;
use axum::body::Body;
use axum::response::sse::Event;
use bsnext_input::route::{DirRoute, ProxyRoute, RouteKind};
use bsnext_resp::response_modifications_layer;
use bytes::Bytes;
use http::StatusCode;
use http_body_util::BodyExt;
use std::sync::Arc;
use std::time::Duration;

use tokio_stream::StreamExt;
use tower::ServiceExt;

use crate::common_layers::add_route_layers;
use tracing::{span, Level};

// use futures_util::stream::{self, Stream};

pub async fn raw_loader(
    State(app): State<Arc<ServerState>>,
    req: Request,
    next: Next,
) -> impl IntoResponse {
    let span = span!(parent: None, Level::INFO, "raw_loader", path = req.uri().path());
    let _guard = span.enter();

    let routes = app.routes.read().await;
    let mut temp_router = matchit::Router::new();
    // let mut app = Router::new();

    // which route kinds can be hard-matched first
    for route in routes.iter() {
        let path = route.path();
        match &route.kind {
            RouteKind::Html { .. }
            | RouteKind::Json { .. }
            | RouteKind::Raw { .. }
            | RouteKind::Sse { .. } => {
                let existing = temp_router.at_mut(path);
                if let Ok(prev) = existing {
                    *prev.value = route.clone();
                    tracing::trace!("updated mutable route at {}", path)
                } else if let Err(err) = existing {
                    match temp_router.insert(path, route.clone()) {
                        Ok(_) => {
                            tracing::trace!(path, "+")
                        }
                        Err(_) => {
                            tracing::error!("❌ {:?}", err.to_string())
                        }
                    }
                }
            }
            _ => {}
        }
    }

    drop(routes);

    let matched = temp_router.at(req.uri().path());

    let Ok(matched) = matched else {
        return next.run(req).await;
    };

    let route = matched.value;
    let _params = matched.params;

    let mut app = Router::new();
    match &route.kind {
        RouteKind::Sse { sse } => {
            let raw = sse.to_owned();
            app = app.route(
                req.uri().path(),
                any(|| async move {
                    let l = raw
                        .lines()
                        .map(|l| l.to_owned())
                        .map(|l| l.strip_prefix("data:").unwrap_or(&l).to_owned())
                        .filter(|l| !l.trim().is_empty())
                        .collect::<Vec<_>>();

                    tracing::trace!(lines.count = l.len(), "sending EventStream");

                    let stream = tokio_stream::iter(l)
                        .throttle(Duration::from_millis(500))
                        .map(|chu| Event::default().data(chu))
                        .map(Ok::<_, Infallible>);

                    Sse::new(stream)
                }),
            )
        }
        RouteKind::Raw { raw } => {
            tracing::trace!("-> served Route::Raw {} {} bytes", route.path, raw.len());
            let moved = raw.clone();
            let p = req.uri().path().to_owned();
            app = app.route(
                req.uri().path(),
                any(|| async move { text_asset_response(&p, &moved) }),
            );
        }
        RouteKind::Html { html } => {
            tracing::trace!("-> served Route::Html {} {} bytes", route.path, html.len());
            let moved = html.clone();
            app = app.route(req.uri().path(), any(|| async { Html(moved) }));
        }
        RouteKind::Json { json } => {
            tracing::trace!("-> served Route::Json {} {}", route.path, json);
            let moved = json.to_owned();
            app = app.route(req.uri().path(), any(|| async move { Json(moved) }));
        }
        RouteKind::Dir(DirRoute { dir: _ }) => {
            unreachable!("should never reach RouteKind::Dir")
        }
        RouteKind::Proxy(ProxyRoute { proxy: _ }) => {
            unreachable!("should never reach RouteKind::Proxy")
        }
    }

    app = add_route_layers(app, route);

    app.layer(middleware::from_fn(tag_raw))
        .layer(middleware::from_fn(response_modifications_layer))
        .oneshot(req)
        .await
        .into_response()
}

async fn tag_raw(req: Request, next: Next) -> Result<impl IntoResponse, (StatusCode, String)> {
    let (mut parts, body) = next.run(req).await.into_parts();
    if parts.status.as_u16() == 200 {
        parts.extensions.insert(MetaData::ServedRaw);
    }
    Ok(Response::from_parts(parts, body))
}

#[allow(dead_code)]
async fn print_request_response(
    req: Request,
    next: Next,
) -> Result<impl IntoResponse, (StatusCode, String)> {
    let res = next.run(req).await;

    let (parts, body) = res.into_parts();
    let bytes = buffer_and_print("response", body).await?;
    let lines = bytes.chunks(150).map(|c| c.to_owned()).collect::<Vec<_>>();

    let stream = tokio_stream::iter(lines)
        .throttle(Duration::from_millis(500))
        .map(Ok::<_, Infallible>);

    let res = Response::from_parts(parts, Body::from_stream(stream));

    Ok(res)
}

async fn buffer_and_print<B>(direction: &str, body: B) -> Result<Bytes, (StatusCode, String)>
where
    B: axum::body::HttpBody<Data = Bytes>,
    B::Error: std::fmt::Display,
{
    let bytes = match body.collect().await {
        Ok(collected) => collected.to_bytes(),
        Err(err) => {
            return Err((
                StatusCode::BAD_REQUEST,
                format!("failed to read {direction} body: {err}"),
            ));
        }
    };

    if let Ok(body) = std::str::from_utf8(&bytes) {
        tracing::debug!("▶️ {direction} body = {body:?}");
    }

    Ok(bytes)
}

pub fn text_asset_response(path: &str, css: &str) -> Response {
    let mime = mime_guess::from_path(path);
    let aas_str = mime.first_or_text_plain();
    let cloned = css.to_owned();
    ([(CONTENT_TYPE, aas_str.to_string())], cloned).into_response()
}
