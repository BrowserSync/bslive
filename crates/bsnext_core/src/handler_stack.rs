use crate::common_layers::add_route_layers;
use crate::handlers::proxy::{proxy_handler, ProxyConfig};
use crate::raw_loader::serve_raw_one;
use crate::serve_dir::try_many_services_dir;
use axum::handler::Handler;
use axum::middleware::from_fn_with_state;
use axum::routing::{any, any_service, get_service, MethodRouter};
use axum::{Extension, Router};
use bsnext_input::route::{DirRoute, Opts, ProxyRoute, RawRoute, Route, RouteKind};
use bsnext_resp::path_matcher::PathMatcher::Def;
use std::collections::HashMap;
use tower_http::services::ServeDir;

#[derive(Debug, PartialEq)]
pub enum HandlerStack {
    None,
    // todo: make this a separate thing
    Raw { raw: RawRoute, opts: Opts },
    Dirs(Vec<DirRouteOpts>),
    Proxy { proxy: ProxyRoute, opts: Opts },
    DirsProxy(Vec<DirRouteOpts>, ProxyRoute),
}

#[derive(Debug, PartialEq)]
pub struct DirRouteOpts {
    dir_route: DirRoute,
    opts: Opts,
}

impl DirRouteOpts {
    pub fn as_serve_dir(&self) -> ServeDir {
        match &self.dir_route.base {
            Some(base_dir) => {
                tracing::trace!(
                    "combining root: `{}` with given path: `{}`",
                    base_dir.display(),
                    self.dir_route.dir
                );
                ServeDir::new(base_dir.join(&self.dir_route.dir))
            }
            None => {
                tracing::trace!("no root given, using `{}` directly", self.dir_route.dir);
                ServeDir::new(&self.dir_route.dir)
            }
        }
    }
}

impl DirRouteOpts {
    fn new(p0: DirRoute, p1: Opts) -> DirRouteOpts {
        Self {
            dir_route: p0,
            opts: p1,
        }
    }
}

pub struct RouteMap {
    pub mapping: HashMap<String, Vec<Route>>,
}

impl RouteMap {
    pub fn new_from_routes(routes: &[Route]) -> Self {
        Self {
            mapping: routes
                .iter()
                .fold(HashMap::<String, Vec<Route>>::new(), |mut acc, route| {
                    acc.entry(route.path.clone())
                        .and_modify(|acc| acc.push(route.clone()))
                        .or_insert(vec![route.clone()]);
                    acc
                }),
        }
    }

    pub fn into_router(self) -> Router {
        let mut router = Router::new();

        tracing::trace!("processing `{}` different routes", self.mapping.len());

        for (path, route_list) in self.mapping {
            tracing::trace!(
                "processing path: `{}` with `{}` routes",
                path,
                route_list.len()
            );

            let stack = routes_to_stack(&route_list);
            let path_router = stack_to_router(&path, stack);

            tracing::trace!("will merge router at path: `{path}`");
            router = router.merge(path_router);
        }

        router
    }
}

pub fn append_stack(state: HandlerStack, route: Route) -> HandlerStack {
    match state {
        HandlerStack::None => match route.kind {
            RouteKind::Raw(raw_route) => HandlerStack::Raw {
                raw: raw_route,
                opts: route.opts,
            },
            RouteKind::Proxy(new_proxy_route) => HandlerStack::Proxy {
                proxy: new_proxy_route,
                opts: route.opts,
            },
            RouteKind::Dir(dir) => HandlerStack::Dirs(vec![DirRouteOpts::new(dir, route.opts)]),
        },
        HandlerStack::Raw { raw, opts } => match route.kind {
            // if a second 'raw' is seen, just use it, discarding the previous
            RouteKind::Raw(raw_route) => HandlerStack::Raw {
                raw: raw_route,
                opts: route.opts,
            },
            // 'raw' handlers never get updated
            _ => HandlerStack::Raw { raw, opts },
        },
        HandlerStack::Dirs(mut dirs) => match route.kind {
            RouteKind::Dir(next_dir) => {
                dirs.push(DirRouteOpts::new(next_dir, route.opts));
                HandlerStack::Dirs(dirs)
            }
            RouteKind::Proxy(proxy) => HandlerStack::DirsProxy(dirs, proxy),
            _ => HandlerStack::Dirs(dirs),
        },
        HandlerStack::Proxy { proxy, opts } => match route.kind {
            RouteKind::Dir(dir) => {
                HandlerStack::DirsProxy(vec![DirRouteOpts::new(dir, route.opts)], proxy)
            }
            _ => HandlerStack::Proxy { proxy, opts },
        },
        HandlerStack::DirsProxy(mut dirs, proxy) => match route.kind {
            RouteKind::Dir(dir) => {
                dirs.push(DirRouteOpts::new(dir, route.opts));
                HandlerStack::DirsProxy(dirs, proxy)
            }
            // todo(alpha): how to handle multiple proxies? should it just override for now?
            _ => HandlerStack::DirsProxy(dirs, proxy),
        },
    }
}

pub fn routes_to_stack(routes: &[Route]) -> HandlerStack {
    routes.iter().fold(HandlerStack::None, |s, route| {
        append_stack(s, route.clone())
    })
}

pub fn stack_to_router(path: &str, stack: HandlerStack) -> Router {
    match stack {
        HandlerStack::None => unreachable!(),
        HandlerStack::Raw { raw, opts } => {
            let svc = any_service(serve_raw_one.with_state(raw));
            Router::new().route_service(path, add_route_layers(svc, path, &opts))
        }
        HandlerStack::Dirs(dirs) => {
            Router::new().nest_service(path, serve_dir_layer(path, &dirs, Router::new()))
        }
        HandlerStack::Proxy { proxy, opts } => {
            let proxy_config = ProxyConfig {
                target: proxy.proxy.clone(),
                path: path.to_string(),
            };
            let router = any(proxy_handler).layer(Extension(proxy_config.clone()));
            Router::new().nest_service(path, add_route_layers(router, path, &opts))
        }
        HandlerStack::DirsProxy(dir_list, proxy) => {
            let r2 = stack_to_router(
                path,
                HandlerStack::Proxy {
                    proxy,
                    opts: Default::default(),
                },
            );
            let r1 = serve_dir_layer(path, &dir_list, Router::new().fallback_service(r2));
            Router::new().nest_service(path, r1)
        }
    }
}

fn serve_dir_layer(path: &str, dir_list_with_opts: &[DirRouteOpts], initial: Router) -> Router {
    let serve_dir_items = dir_list_with_opts
        .iter()
        .map(|dir_route| {
            let src = dir_route.as_serve_dir();
            let service = get_service(src);
            add_route_layers(service, path, &dir_route.opts)
        })
        .collect::<Vec<MethodRouter>>();

    initial.layer(from_fn_with_state(
        (path.to_string(), serve_dir_items),
        try_many_services_dir,
    ))
}

#[cfg(test)]
mod test {
    use super::*;

    use crate::server::router::common::to_resp_parts_and_body;
    use axum::body::Body;
    use bsnext_input::Input;
    use http::Request;
    use insta::assert_debug_snapshot;
    use tower::ServiceExt;

    #[test]
    fn test_handler_stack_01() -> anyhow::Result<()> {
        let yaml = include_str!("../../../examples/basic/handler_stack.yml");
        let input = serde_yaml::from_str::<Input>(&yaml)?;
        let first = input
            .servers
            .iter()
            .find(|x| x.identity.is_named("raw"))
            .unwrap();

        let actual = routes_to_stack(&first.routes);
        assert_debug_snapshot!(actual);
        Ok(())
    }

    #[test]
    fn test_handler_stack_02() -> anyhow::Result<()> {
        let yaml = include_str!("../../../examples/basic/handler_stack.yml");
        let input = serde_yaml::from_str::<Input>(&yaml)?;
        let first = input
            .servers
            .iter()
            .find(|x| x.identity.is_named("2dirs+proxy"))
            .unwrap();

        let actual = routes_to_stack(&first.routes);

        assert_debug_snapshot!(actual);
        Ok(())
    }
    #[test]
    fn test_handler_stack_03() -> anyhow::Result<()> {
        let yaml = include_str!("../../../examples/basic/handler_stack.yml");
        let input = serde_yaml::from_str::<Input>(&yaml)?;
        let first = input
            .servers
            .iter()
            .find(|s| s.identity.is_named("raw+opts"))
            .unwrap();

        let actual = routes_to_stack(&first.routes);

        assert_debug_snapshot!(actual);
        Ok(())
    }

    #[tokio::test]
    async fn test_routes_to_router() -> anyhow::Result<()> {
        let yaml = include_str!("../../../examples/basic/handler_stack.yml");
        let input = serde_yaml::from_str::<Input>(&yaml)?;

        {
            let first = input
                .servers
                .iter()
                .find(|x| x.identity.is_named("raw"))
                .unwrap();

            let route_map = RouteMap::new_from_routes(&first.routes);
            let router = route_map.into_router();
            let request = Request::get("/styles.css").body(Body::empty())?;

            // Define the request
            // Make a one-shot request on the router
            let response = router.oneshot(request).await?;
            let (_parts, body) = to_resp_parts_and_body(response).await;

            assert_eq!(body, "body { background: red }");
        }

        Ok(())
    }
}
