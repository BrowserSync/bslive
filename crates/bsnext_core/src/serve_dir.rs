use axum::body::Body;
use axum::extract::{Request, State};
use axum::middleware::Next;
use axum::response::IntoResponse;
use axum::routing::MethodRouter;
use http::{StatusCode, Uri};
use tower::ServiceExt;

pub async fn try_many_services_dir(
    State(router_list): State<Vec<MethodRouter>>,
    uri: Uri,
    req: Request,
    next: Next,
) -> impl IntoResponse {
    tracing::trace!(?uri, "{} services", router_list.len());

    let (a, b) = req.into_parts();

    for (index, method_router) in router_list.into_iter().enumerate() {
        // tracing::trace!(?path_buf);
        let req_clone = Request::from_parts(a.clone(), Body::empty());
        let result = method_router.oneshot(req_clone).await;
        match result {
            Ok(result) if result.status() == 404 => {
                tracing::trace!("  ❌ not found at index {}, trying another", index,);
                continue;
            }
            Ok(result) if result.status() == 405 => {
                tracing::trace!("  ❌ 405, trying another...");
                continue;
            }
            Ok(result) => {
                tracing::trace!(
                    ?index,
                    " - ✅ a non-404 response was given {}",
                    result.status()
                );
                return result.into_response();
            }
            Err(e) => {
                tracing::error!(?e);
                return StatusCode::INTERNAL_SERVER_ERROR.into_response();
            }
        }
    }
    tracing::trace!(" - REQUEST was NOT HANDLED BY SERVE_DIR (will be sent onwards)");
    let r = Request::from_parts(a.clone(), b);
    next.run(r).await
    // StatusCode::NOT_FOUND.into_response()
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::handler_stack::RouteMap;
    use crate::server::router::common::to_resp_parts_and_body;

    use crate::runtime_ctx::RuntimeCtx;
    use bsnext_input::route::Route;
    use std::env::current_dir;

    #[tokio::test]
    async fn test() -> anyhow::Result<()> {
        let current = current_dir()?;
        let parent = current.parent().unwrap().parent().unwrap().to_owned();

        let routes_input = format!(
            r#"
            - path: /
              dir: examples/basic/public
              base: {base}
            - path: /
              dir: examples/kitchen-sink
              base: {base}
        "#,
            base = parent.display()
        );

        let routes = serde_yaml::from_str::<Vec<Route>>(&routes_input)?;

        {
            let router = RouteMap::new_from_routes(&routes).into_router(&RuntimeCtx::default());
            let expected_body = include_str!("../../../examples/basic/public/index.html");

            // Define the request
            let request = Request::get("/index.html").body(Body::empty())?;
            // Make a one-shot request on the router
            let response = router.oneshot(request).await?;
            let (_parts, actual_body) = to_resp_parts_and_body(response).await;
            assert_eq!(actual_body, expected_body);
        }

        {
            let router = RouteMap::new_from_routes(&routes).into_router(&RuntimeCtx::default());
            let expected_body = include_str!("../../../examples/kitchen-sink/input.html");

            // Define the request
            let request = Request::get("/input.html").body(Body::empty())?;
            // Make a one-shot request on the router
            let response = router.oneshot(request).await?;
            let (_parts, actual_body) = to_resp_parts_and_body(response).await;
            assert_eq!(actual_body, expected_body);
        }

        Ok(())
    }
}
