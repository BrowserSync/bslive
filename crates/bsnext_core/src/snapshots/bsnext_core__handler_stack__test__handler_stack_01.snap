---
source: crates/bsnext_core/src/handler_stack.rs
expression: actual
---
Raw {
    raw: Raw {
        raw: "body { background: red }",
    },
    opts: Opts {
        cors: None,
        delay: None,
        watch: Bool(
            true,
        ),
        inject: Bool(
            true,
        ),
        headers: None,
        compression: None,
    },
}
